implement inherent str constructors

m4rch3n1ng opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

implement #131114

this implements
- str::from_utf8
- str::from_utf8_mut
- str::from_utf8_unchecked
- str::from_utf8_unchecked_mut

i left `std::str::from_raw_parts` and `std::str::from_raw_parts_mut` out of this as those are unstable and were not mentioned by the tracking issue or the original pull request, but i can  add those here as well.

i was also unsure of what to do with the `rustc_const_(un)stable` attributes: i removed the `#[rustc_const_stable]` attribute from `str::from_utf8`, `str::from_utf8_unchecked` and `str::from_utf8_unchecked_mut`, and left the`#[rust_const_unstable]` in `str::from_utf8_mut` (btw why is that one not const stable yet with #57349 merged?).

is there a way to redirect users to the stable `std::str::from_utf8` instead of only saying "hey this is unstable"?

for now i just removed the check for `str::from_utf8` in the test in `tests/ui/suggestions/suggest-std-when-using-type.rs`.

m4rch3n1ng committed:
implement inherent str constructors

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13124504814/job/36617943798) [(plain)](https://github.com/rust-lang/rust/commit/c702c6833607930cf0839539bf0a0b3908f64e63/checks/36617943798/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling compiler_builtins v0.1.143
error: `slice` is both a module and a primitive type
   --> library/core/src/str/mod.rs:172:22
    |
172 |     /// [byteslice]: slice
    |                      ^^^^^ ambiguous link
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
help: to link to the module, prefix with `mod@`
    |
    |
172 |     /// [byteslice]: mod@slice
help: to link to the primitive type, prefix with `prim@`
    |
    |
172 |     /// [byteslice]: prim@slice

error: could not document `core`
warning: build failed, waiting for other jobs to finish...
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Mon Feb  3 23:28:31 UTC 2025
  network time: Mon, 03 Feb 2025 23:28:31 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

m4rch3n1ng committed:
specify a prim@slice in docs

i am not quite sure how this failure is in any way related to this pr,
since i am only touching inherent functions on str? but sure.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13125129880/job/36619807895) [(plain)](https://github.com/rust-lang/rust/commit/9cedf45aff1a53b8ceecb8f9965c6d75cbe6a48b/checks/36619807895/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 27.9s done
#22 DONE 31.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.98s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
warning: target feature `x87` must be enabled to ensure that the ABI of the current target can be implemented correctly
  |
  = note: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #116344 <https://github.com/rust-lang/rust/issues/116344>
---
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.14s
##[endgroup]
##[group]Testing stage0 Linkcheck (x86_64-unknown-linux-gnu)
std/primitive.str.html:77: broken link - `/checkout/obj/build/x86_64-unknown-linux-gnu/std/string/struct.String.html`
core/primitive.str.html:77: broken link - `/checkout/obj/build/x86_64-unknown-linux-gnu/std/string/struct.String.html`
number of HTML files scanned: 44204
number of HTML redirects found: 13804
number of links checked: 2426749
number of links ignored due to external: 91817
number of links ignored due to external: 91817
number of links ignored due to exceptions: 24
number of intra doc links ignored: 8
errors found: 2
found some broken links
NOTE: if you are adding or renaming a markdown file in a mdBook, don't forget to register the page in SUMMARY.md
Command has failed. Rerun with -v to see more details.
  local time: Tue Feb  4 00:24:18 UTC 2025
  network time: Tue, 04 Feb 2025 00:24:18 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jhpratt commented:
> i left std::str::from_raw_parts and std::str::from_raw_parts_mut out of this as those are unstable and were not mentioned by the tracking issue or the original pull request, but i can add those here as well.

That can be done in a separate PR if you wish!

> i was also unsure of what to do with the rustc_const_(un)stable attributes: i removed the #[rustc_const_stable] attribute from str::from_utf8, str::from_utf8_unchecked and str::from_utf8_unchecked_mut, and left the#[rust_const_unstable] in str::from_utf8_mut (btw why is that one not const stable yet with https://github.com/rust-lang/rust/issues/57349 merged?).

The attributes are correct. The compiler will (thankfully!) complain if they're _clearly_ wrong, but there are edge cases that reviewers can keep an eye out for if you're ever not sure. As to why the mutable version is not const stable, that's a great question. I asked in the tracking issue, so the team should look at it soon-ish.

> is there a way to redirect users to the stable std::str::from_utf8 instead of only saying "hey this is unstable"?

Probably, but don't worry about it. I'll ping @estebank to see if there's anything that can be done here (ideally for the general case, as there's a number of things like this).

---

Thank you for doing this! It looks good to me, so merging :slightly_smiling_face: 

@bors r+

bors commented:
:pushpin: Commit 15adc38ffc4608855c959ec521fe0124e1dffb2c has been approved by `jhpratt`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jhpratt 15adc38ffc4608855c959ec521fe0124e1dffb2c -->
<!-- homu: {"type":"Approved","sha":"15adc38ffc4608855c959ec521fe0124e1dffb2c","approver":"jhpratt","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
